### ISSUE TITLE
cobrapy hangs when GLPK can't find a solution under model.optimize(solver="glpk")

### ISSUE THREAD

**Issue by melomcr** [opencobra/cobrapy#117]
When running a _very_ large number of simulations using different upper bounds for reactions in the model, some times GLPK cannot find a solution. In the rare event is happens (one every few hundred thousand cases) cobrapy hangs and crashes some time later. It took me a while to find the issue because the only error output would be from MPI saying one of the ranks terminated unexpectedly. Using C-GLPK, this rare event produces an exception (the output is copied below), so using a try-except I could treat such cases. I am not familiar with the inner workings of GLPK, but I believe some similar treatment should be creaded to handle such cases.

Traceback (most recent call last):
  File "/home/melomcr/cobrapy/cobrapy_mpi_v2.py", line 650, in <module>
    resultVmaxGR = prepare_and_optimize_model(SimPerCore, model, proteins, v_info_dict)
  File "/home/melomcr/cobrapy/cobrapy_mpi_v2.py", line 315, in prepare_and_optimize_model
    localModel.optimize(solver='cglpk')  # 'gurobi' or 'cglpk' or 'glpk'
  File "/home/melomcr/Programas/cobrapy-master/cobra/core/Model.py", line 301, in optimize
    *_kwargs)
  File "/home/melomcr/Programas/cobrapy-master/cobra/solvers/**init**.py", line 110, in optimize
    the_solution = solver_function.solve(cobra_model, *_kwargs)
  File "cglpk.pyx", line 186, in cobra.solvers.cglpk.GLP.solve (cobra/solvers/cglpk.c:3392)
  File "cglpk.pyx", line 170, in cobra.solvers.cglpk.GLP.solve_problem (cobra/solvers/cglpk.c:3155)
AssertionError

**Comment by LucasYEAST** [comment]
I still encounter this issue. When I run many pFBA simulations changing upper and lower bounds, every 500 simulations or so the solver hangs without throwing an error. I believe it is not possible anymore to specify cglpk instead of glpk. I'm using Cobra 0.11.3 with only the glpk and glpk-exact solvers available.

**Comment by cdiener** [comment]
Hi, sorry for the delayed reply. So I suspect that is actaully not the same error as initially reported in the PR since that was something really specific to glpk 4.50. 

There are some parts of your code I did not get, for instance `except Infeasible:`. I don't see your imports but `Infeasible` is not really declared anywhere and the `INFEASIBLE` from optlang is a string and not an Exception object. So that except should never trigger but give you some error. If you want to catch the error thrown by cobrapy you would need to use `cobra.OptimizationError`. By any chance are you running that code in parallel using `multiprocessing.Pool` or something similar? A common problem is not calling `join` on processes which makes you process run forever without passing the error to the main process.

**Comment by Midnighter** [comment]
Side note: `Infeasible` is an exception that we define in cobrapy. (We're not being very systematic or exhaustive there so I think a general `OptimizationError` with appropriate messages would be better.)

**Comment by cdiener** [comment]
You are right, changed my comment. But everything that uses `get_solution` internally should always throw an `OptimizationError`.

**Comment by LucasYEAST** [comment]
Thanks for your reply @cdiener . The issue was solved by using Gurobi instead of GLPK. 
While using GLPK, I've never seen the OptimizationError, it always throws an 'Infeasible' error. When I changed to GUROBI I added OptimizationError to the exception, I'll try and run the model again with GLPK to see whether it still hangs after including the OptimizationError.
``` 
from cobra.exceptions import Infeasible, OptimizationError
...
try:
    if analysis == 'pfba':
        sol = pfba(m).to_frame().fluxes
        death = False #Solution is feasible
        if return_model:
            return m, death
        else:
            sol.to_msgpack(path_or_buf = "input_files/"+path+ "/fluxdistributions/" + ternary_string + ".msgpack")
except (Infeasible,OptimizationError):  
    sol = pd.Series({False:False})
    death = True
...
```

I'm not running anything in parallel.

**Comment by Midnighter** [comment]
`Infeasible` inherits from `OptimizationError` so with `OptimizationError` you're on the save side and everything should be caught. Please let us know what happens with GLPK.

**Comment by LucasYEAST** [comment]
Running the model with GLPK while using both 'Infeasible'  and 'OptimizationError', did not catch the error. The model still hangs, this time after running 1325 solutions. Notably, it has always been the case that the program is insensitive to interrupting the kernel when it is hanging (I'm running from Jupyter Notebook).

**Comment by cdiener** [comment]
@LucasYEAST copbrapy does not use the systems GLPK anymore but comes with its own version. You can use the following code to figure out the GLPK version cobrapy uses:

```Python
In [1]: import swiglpk

In [2]: swiglpk.glp_version()
Out[2]: '4.61'
```

Compiled extensions are pretty much insensitive to the kernel interruption so that makes sense. Unfortunately it is pretty hard do diagnose where the problem is. A minimal reproducible examples would be great but might be difficult here. One thing you could try is resetting the solver by hand every once in a while. Here is a quick code fragment:

```Python
from swiglpk import glp_adv_basis

#... your code
# place where you want to reset the solver
glp_adv_basis(model.solver.problem, 0)
# ... rest of your code
```

If that works it means the solver gets stuck in an ill-conditioned solution basis. I have never seen that happen for models with less than 20,000 reactions but you never know...

**Comment by LucasYEAST** [comment]
@cdiener I just ran the model while resetting before every new solution was created. I've added the following code to the start of the function that creates a solution:
```
if type(m.solver) == optlang.glpk_interface.
    glp_adv_basis(m.solver.problem, 0)
    print "solver reset"
```
While importing:
```
from swiglpk import glp_adv_basis
import optlang.glpk_interface
```

The solver however still hangs.

My glp version is 4.60

**Comment by LucasYEAST** [comment]
Updated both:

pip optlang version: 1.4.1
pip swiglpk version: 1.4.4

Some strange things happened:
After updating optlang it was not longer possible to use 'gurobi' as a solver:
```
import cobra
m = cobra.io.read_sbml_model("models/yeast_7.6_cobra.xml")
m.solver = 'gurobi'

>> AttributeError                            Traceback (most recent call last)
>> <ipython-input-11-4bc87e47cbee> in <module>()
>>      1 m = cobra.io.read_sbml_model("models/yeast_7.6_cobra.xml")
>> ----> 2 m.solver = 'gurobi'
...
>> AttributeError: 'Variable' object has no attribute 'problem'
```
Downgrading back to optlang version 1.4.0 solved the issue.

Second, my glp_version() is still 4.60:
```
import swiglpk
swiglpk.glp_version()
>> '4.60'
```

**Comment by Midnighter** [comment]
This is a known bug with optlang and the next relase should fix it. https://github.com/biosustain/optlang/issues/158

**Comment by phantomas1234** [comment]
The fix has been roled out in optlan==1.4.2 https://github.com/biosustain/optlang/issues/158. So, I am closing this one. Please reopen, if the issue persists.
